osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs

```diff
@@ -485,16 +485,32 @@ private void addControlPoint(double time, ControlPoint point, bool timingChange)
         private void flushPendingPoints()
         {
             // Changes from non-timing-points are added to the end of the list (see addControlPoint()) and should override any changes from timing-points (added to the start of the list).
+            // As non-timing-points should not set the OmitFirstBarline setting, we will honor the setting of the last effect point, the effect point generated by the timing-point (if any).
+            EffectControlPoint effectPoint = null;
+
             for (int i = pendingControlPoints.Count - 1; i >= 0; i--)
             {
-                var type = pendingControlPoints[i].GetType();
-                if (pendingControlPointTypes.Contains(type))
-                    continue;
-
-                pendingControlPointTypes.Add(type);
-                beatmap.ControlPointInfo.Add(pendingControlPointsTime, pendingControlPoints[i]);
+                if (pendingControlPoints[i] is EffectControlPoint pendingEffectPoint)
+                {
+                    if (effectPoint == null)
+                        effectPoint = pendingEffectPoint;
+                    else
+                        effectPoint.OmitFirstBarLine = pendingEffectPoint.OmitFirstBarLine;
+                }
+                else
```

Comment:
Well okay, maybe it still works because of your manual tracking via the `effectPoint` variable. But I believe the whole method would read better if you remove the reliance on that variable.

I think a better way of writing this would be:

```csharp
        private void flushPendingPoints()
        {
            // Changes from non-timing-points are added to the end of the list (see addControlPoint()) and should override any changes from timing-points (added to the start of the list).

            for (int i = pendingControlPoints.Count - 1; i >= 0; i--)
            {
                var cp = pendingControlPoints[i];
                var type = cp.GetType();

                if (pendingControlPointTypes.Contains(type))
                {
                    // In stable, OmitFirstBarLine could only be set on timing points (ie. non-inherited points).
                    // From this perspective, it seems odd that we have the OmitFirstBarLine setting in EffectControlPoint as this breaks that assumption.
                    //
                    // To make sure legacy beatmaps are handled correctly, we therefore need to make sure to respect the earliest OmitFirstBarLine setting
                    // for the current point in time (aka the one read first from the file).
                    //
                    // The easiest way to handle this is to transfer any non-false values from the current flush group to the actual ControlPointInfo instance.
                    if ((cp as EffectControlPoint)?.OmitFirstBarLine == true)
                    {
                        beatmap.ControlPointInfo.EffectPointAt(cp.Time).OmitFirstBarLine = true;
                    }

                    continue;
                }

                pendingControlPointTypes.Add(type);
                beatmap.ControlPointInfo.Add(pendingControlPointsTime, cp);
            }

            pendingControlPoints.Clear();
            pendingControlPointTypes.Clear();
        }

```

If you'd like to attempt to fix the tests and apply this method then I'll leave it to you, or if you'd prefer I can open a second PR with this and accompanying tests.

Reply:
I'll do so later today.